torchgeo/trainers/simclr.py

```diff
@@ -0,0 +1,274 @@
+
+        grayscale_weights = grayscale_weights or torch.ones(in_channels)
+        self.augmentations = augmentations or simclr_augmentations(
+            size, grayscale_weights
+        )
+
+        # Create backbone
+        self.backbone = timm.create_model(
+            model, in_chans=in_channels, num_classes=0, pretrained=weights is True
+        )
+
+        # Load weights
+        if weights and weights is not True:
+            if isinstance(weights, WeightsEnum):
+                state_dict = weights.get_state_dict(progress=True)
+            elif os.path.exists(weights):
+                _, state_dict = utils.extract_backbone(weights)
+            else:
+                state_dict = get_weight(weights).get_state_dict(progress=True)
+            self.backbone = utils.load_state_dict(self.backbone, state_dict)
+
+        # Create projection head
+        input_dim = self.backbone.num_features
+        if hidden_dim is None:
+            hidden_dim = input_dim
+        if output_dim is None:
+            output_dim = input_dim
+
+        self.projection_head = SimCLRProjectionHead(
+            input_dim, hidden_dim, output_dim, layers
+        )
+
+        # Define loss function
+        self.criterion = NTXentLoss(temperature, memory_bank_size, gather_distributed)
+
+        # Initialize moving average of output
+        self.avg_output_std = 0.0
+
+        # TODO
+        # v1+: add global batch norm
+        # v2: add selective kernels, channel-wise attention mechanism
+
+    def forward(self, x: Tensor) -> tuple[Tensor, Tensor]:
+        """Forward pass of the model.
+
+        Args:
+            x: Mini-batch of images.
+
+        Returns:
+            Output from the backbone and projection head.
+        """
+        h = self.backbone(x)  # shape of batch_size x num_features
+        z = self.projection_head(h)
+        return cast(Tensor, z), cast(Tensor, h)
+
+    def training_step(self, batch: dict[str, Tensor], batch_idx: int) -> Tensor:
+        """Compute the training loss and additional metrics.
+
+        Args:
+            batch: The output of your DataLoader.
+            batch_idx: Integer displaying index of this batch.
+
+        Returns:
+            The loss tensor.
+        """
+        x = batch["image"]
+
+        in_channels = self.hparams["in_channels"]
+        assert x.size(1) == in_channels or x.size(1) == 2 * in_channels
+
+        if x.size(1) == in_channels:
+            x1 = x
+            x2 = x
+        else:
+            x1 = x[:, :in_channels]
+            x2 = x[:, in_channels:]
+
+        with torch.no_grad():
+            x1 = self.augmentations(x1)
+            x2 = self.augmentations(x2)
+
+        z1, h1 = self(x1)
+        z2, h2 = self(x2)
+
+        loss = self.criterion(z1, z2)
+
+        # Calculate the mean normalized standard deviation over features dimensions.
+        # If this is << 1 / sqrt(h1.shape[1]), then the model is not learning anything.
```

Comment:
@calebrob6 do you want this in every SSL trainer or no SSL trainers? Want to be consistent.

Reply:
It is a good thing to monitor in training per the SimSiam paper so we shouldn't remove it just for convenience, however it also isn't urgent to add it to other trainers

Reply:
I'll add it to MoCo, should only take a minute.